keps/0017-pipe-tasks.md

```diff
@@ -0,0 +1,122 @@
+
+### Non-Goals
+
+Allowing to pipe all kind of files (>1Mb) between tasks requires a general-purpose storage solution (e.g. S3) available in the cluster. While certainly possible, this is not the goal of this KEP.
+
+## Proposal
+
+This section describes how pipe tasks and files they produce are configured in the operator. Here is a task definition that produces a file that will be stored as a Secret, referenced by `{{.Pipes.Certificate}}` key:
+```yaml
+steps:
+    - name: bootstrap
+    tasks:
+        - gencert
+            - pipe:
+                - file: /usr/share/MyKey.key
+                  kind: Secret # ConfigMap
+                  key: {{.Pipes.Certificate}}
+``` 
+The key `{{.Pipes.Certificate}}` value will be generated by KUDO to avoid collisions and to make sure that it is not used before the file is generated. In the above example we would generate a secret name like `deploy.bootstrap.gencert.pipes.certificate-#hash` to capture which plan/phase/step/task it belongs too as well as the hash of its content. KUDO would use this hash to avoid recreating secrets for unchanged files on a plan rerun. We would also use labels ensuring that the secret is cleaned up when the corresponding Instance is removed.
+
+The corresponding `gencert` task can only have one resource definition:
+```yaml
+tasks:
+  gencert:
+    resources:
+      - gencert.yaml
+```
+
+For the pipe task resources, we only allow a [ContainerSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#container-v1-core) to be specified. Reasons for that are explained below in the implementation details. The ContainerSpec has to define a shared volume where the files are stored.
+
+```yaml
+spec:
+  volumes:
+  - name: shared-data
+    emptyDir: {}
+
+  containers:
+    - name: gencert
+      image: frapsoft/openssl
+      imagePullPolicy: Always
+      command: [ "sh", "-c" ]
+      args:
+        - openssl req -new -newkey rsa:4096 -x509 -sha256 -days 365 -nodes -out MyCertificate.crt -keyout /usr/share/MyKey.key
+      volumeMounts:
+        - name: shared-data
+          mountPath: /usr/share/
+```
+
+Any subsequent task resource might reference previously generated file by its key `{{.Pipes.Certificate}}` e.g.:
+```yaml
+spec:
+  containers:
+  - name: myapp
+    image: myapp:latest
+    volumeMounts:
+    - name: cert
+      mountPath: /еtc/certs
+  volumes:
+  - name: cert
+    secret:
+      secretName: {{.Pipes.Certificate}}
+```
+
+### Limitations
+- File generating container has to be idempotent as the container might be executed multiple times on failure. If that's not the case, `restartPolicy: Never` has to be set which would prevent the task from finishing successfully.
```

Comment:
what happens if this container fail? Does it basically cancel execution of the whole plan?

Reply:
I think it has to 🤷‍♀ maybe note here that actually using the restartpolicy of never mean that the whole plan will be fail. Then it will be retried as a whole? 🤔

Reply:
This means the current resource in the step will not get [healthy](https://github.com/kudobuilder/kudo/blob/master/pkg/controller/planexecution/plan_execution_engine.go#L168) so the same rules would apply as for any other type of resource.